### DISCUSSION TITLE
TS error when creating a variant (discrimated union) dynamically

### DISCUSSION THREAD

**Discussion by NixBiks** [open-circle/valibot#569] [Q&A] [answered]
I'm trying to create a variant schema from an existing object but I get a TS error. Here is a minimal example

```ts
import * as v from 'valibot';


const attributes = {
  ORTH: {
    validation: v.string(),
    description: "The exact text of a token.",
  },
  IS_ALPHA: {
    validation: v.boolean(),
    description: "Is the token an alpha character?",
  },
} as const;

const AttributeSchema = v.variant(
  "key",
  // ts error here
  Object.entries(attributes).map(([key, { validation }]) =>
    v.object({
      key: v.literal(key),
      value: validation,
    })
  )
);

const result = v.safeParse(AttributeSchema, { key: "ORTH", value: "21" });
console.log(result);
```

The schema parses fine but I get a TS error in the second argument of `v.variant`.

```
No overload matches this call.
  Overload 1 of 2, '(key: "key", options: VariantOptions<"key">, pipe?: Pipe<any> | undefined): VariantSchema<"key", VariantOptions<"key">, any>', gave the following error.
    Argument of type 'ObjectSchema<{ key: LiteralSchema<"ORTH" | "IS_ALPHA", "ORTH" | "IS_ALPHA">; value: StringSchema<string> | BooleanSchema<boolean>; }, undefined, { ...; }>[]' is not assignable to parameter of type 'VariantOptions<"key">'.
      Source provides no match for required element at position 0 in target.
  Overload 2 of 2, '(key: "key", options: VariantOptions<"key">, message?: ErrorMessage | undefined, pipe?: Pipe<any> | undefined): VariantSchema<"key", VariantOptions<"key">, any>', gave the following error.
    Argument of type 'ObjectSchema<{ key: LiteralSchema<"ORTH" | "IS_ALPHA", "ORTH" | "IS_ALPHA">; value: StringSchema<string> | BooleanSchema<boolean>; }, undefined, { ...; }>[]' is not assignable to parameter of type 'VariantOptions<"key">'.
```

Any TS hero that can help me out here?

**Comment by fabian-hiller** [comment, marked as answer]
The problem is that `variant` currently requires that you specify at least two options (otherwise `variant` makes no sense), but TS is not smart enough to know that the return type of `Object.entries(attributes).map(...)` will have two elements. It is just typed as a generic array of object schemas instead of a tuple.

I can consider changing the types of `variant` so that TS does not complain. But this has two drawbacks. The first is that TS will stop complaining if you define `variant` with no or only one option. The second is that it will require us to add additional code that slightly increases the bundle size to return an issue if no option (empty array) is specified.

**Reply by NixBiks** [reply]
Ah. Just leave it. I can just ts-ignore or make a workaround by extracting the first two elements and explicitly showing that there are at least two elements. 

Thanks for answering and for Valibot 🙏🏼

**Reply by fabian-hiller** [reply]
For our ongoing rewrite in #502, I changed it to a normal array. I assume that flexibility is a bit more important than a TS error for variants with no or a single option.
